src/syrupy/session.py

```diff
@@ -93,6 +101,27 @@ def flush_snapshot_write_queue(self) -> None:
                 )
         self._queued_snapshot_writes = {}
 
+    def recall_snapshot(
+        self,
+        extension: "AbstractSyrupyExtension",
+        test_location: "PyTestLocation",
+        index: "SnapshotIndex",
+    ) -> Optional["SerializedData"]:
+        """Find the current value of the snapshot, for this session, either a pending write or the actual snapshot."""
+
+        key = self._snapshot_write_queue_key(extension, test_location, index)
+        queue = self._queued_snapshot_writes.get(key)
+        if queue:
+            # find the last (i.e. most recent) write to this index/location in the queue:
+            for queue_data, queue_test_location, queue_index in reversed(queue):
+                if queue_index == index and queue_test_location == test_location:
+                    return queue_data
```

Comment:
This loops is a potential performance issue. In particular: it scales like O(length of `queue`), i.e. number of pending writes.

This function is called twice in `SnapshotAssertion._assert`:

- unconditionally: https://github.com/syrupy-project/syrupy/blob/ef8189c68460593fead9c484405b755e272c8cca/src/syrupy/assertion.py#L316
- conditionally, only when diffing: https://github.com/syrupy-project/syrupy/blob/ef8189c68460593fead9c484405b755e272c8cca/src/syrupy/assertion.py#L320

If every assertion in a ambr file needs an update, this leads to quadratic O(number of assertions in that file^2) behaviour:

- the first `assert ... == snapshot` has an empty `queue`, and adds an item
- the second checks one element of the queue (doesn't match), and adds an item
- the third checks two elements of the queue (don't match) and adds an item
- ...
- the Nth checks N - 1 elements of the queue (all don't match) and adds an item

i.e. N assertions ends up iterating over 0 + 1 + 2 + ... + (N - 1) = (N - 1) * (N - 2) / 2 = O(N^2) elements of `queue`.

I think this is probably okay:

- this scales only per snapshot location, not across the full testing session, since each `queue` is just for one key, which incorporates the snapshot location (i.e. if one has 10 assertions in each of 100 test files, this is scaling with N = 10, not N = 10 * 100).
- how often does one have to completely re-write a _whole_ file?
- the work here is individually cheap (i.e. the constant factor is low)

Reply:
If the benchmarking is accurate, there appears to be a bit of a slowdown: https://github.com/syrupy-project/syrupy/pull/927#issuecomment-2563177603

I ran `inv benchmark` locally, and this branch:

```
------------------------------------------------------------------------------ benchmark: 3 tests ------------------------------------------------------------------------------
Name (time in s)         Min               Max              Mean            StdDev            Median               IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_1000x_reads      1.0756 (1.0)      1.2278 (1.0)      1.1648 (1.0)      0.0630 (4.10)     1.1974 (1.0)      0.0948 (3.49)          2;0  0.8585 (1.0)           5           1
test_standard         1.1904 (1.11)     1.3294 (1.08)     1.2822 (1.10)     0.0534 (3.48)     1.2960 (1.08)     0.0439 (1.62)          1;1  0.7799 (0.91)          5           1
test_1000x_writes     1.3974 (1.30)     1.4278 (1.16)     1.4089 (1.21)     0.0153 (1.0)      1.3986 (1.17)     0.0272 (1.0)           1;0  0.7098 (0.83)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

and "main":

```
------------------------------------------------------------------------------ benchmark: 3 tests ------------------------------------------------------------------------------
Name (time in s)         Min               Max              Mean            StdDev            Median               IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_1000x_reads      1.0264 (1.0)      1.1148 (1.0)      1.0855 (1.0)      0.0362 (1.0)      1.1006 (1.0)      0.0474 (1.0)           1;0  0.9212 (1.0)           5           1
test_standard         1.0904 (1.06)     1.1974 (1.07)     1.1461 (1.06)     0.0477 (1.32)     1.1414 (1.04)     0.0876 (1.85)          2;0  0.8725 (0.95)          5           1
test_1000x_writes     1.0937 (1.07)     1.2011 (1.08)     1.1406 (1.05)     0.0405 (1.12)     1.1313 (1.03)     0.0534 (1.13)          2;0  0.8767 (0.95)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Reply:
> this scales per snapshot location,

This comes up when users are leveraging test parameterization over data sets. i.e. the benchmark case of parameterizing over 1000 test cases isn't too unrealistic from how I've seen syrupy used in the past.

Reply:
FYI, the benchmarking is broken in CI for contributors (something to do with permissions in the github workflow I need to fix), however you should be able to run it locally with `inv benchmark`.